src/main/java/com/google/sps/model/AuthenticatedHttpServlet.java

```diff
@@ -0,0 +1,103 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.model;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.sps.utility.AuthenticationUtility;
+import java.io.IOException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** HttpServlet that enforces the verification of user tokens */
+public abstract class AuthenticatedHttpServlet extends HttpServlet {
+  // Error message if user is not authenticated
+  protected static final String ERROR_403 = "Authentication tokens not present / invalid";
+
+  /**
+   * Verifies user credentials on GET (sending a 403 error in the case that the user is not properly
+   * authenticated).
+   *
+   * @param request Http request sent from client
+   * @param response Http response to be sent back to the client
+   * @throws IOException if there is an issue processing the request
+   */
+  @Override
+  public final void doGet(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    Credential googleCredential = loadCredential(request, response);
+    if (googleCredential != null) {
+      doGet(request, response, googleCredential);
+    }
+  }
+
+  /**
+   * Verifies user credentials on POST (sending a 403 error in the case that the user is not
+   * properly authenticated).
+   *
+   * @param request Http request sent from client
+   * @param response Http response to be sent back to the client
+   * @throws IOException if there is an issue processing the request
+   */
+  @Override
+  public final void doPost(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    Credential googleCredential = loadCredential(request, response);
+    if (googleCredential != null) {
+      doPost(request, response, googleCredential);
+    }
+  }
+
+  /**
+   * Handle GET request
+   *
+   * @param request HTTP request from client
+   * @param response Http response to be sent to client
+   * @param googleCredential valid, verified google credential object
+   * @throws IOException is there is an issue processing the request
+   */
+  public abstract void doGet(
+      HttpServletRequest request, HttpServletResponse response, Credential googleCredential)
+      throws IOException;
+
+  /**
+   * Handle GET request
+   *
+   * @param request HTTP request from client
+   * @param response Http response to be sent to client
+   * @param googleCredential valid, verified google credential object
+   * @throws IOException is there is an issue processing the request
+   */
+  public abstract void doPost(
```

Comment:
Any reason for this method to be public instead of protected?

Reply:
We should chat offline about this - switching to protected is probably better semantically, but has some implications for unit testing